### PR TITLE
Unit Testing for ViewModel, Data and Domain

### PR DESCRIPTION
**Pull request by bententuan**
Apply unit testing for ViewModel layer, Data layer and Domain layer

### CODE REVIEW

File: app/src/test/java/com/example/tmdb/ui/features/login/LoginViewModelTest.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    @Test
+    fun `login with correct user name and password`() = runTest {
+        viewModel.apply {
+            val params = LoginParams("user01", "1234")
+            updateUsername(params.username)
+            updatePassword(params.password)
+            login()
+            coVerify {
+                loginUseCase(params)
+            }
```
Comment (huuphuoc1396) [inline app/src/test/java/com/example/tmdb/ui/features/login/LoginViewModelTest.kt:44]: We should check on the success that the navigator can trigger correctly or not. Also, when on the error the error event and single event can trigger.

File: app/src/test/java/com/example/tmdb/ui/features/login/LoginViewModelTest.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+@ExperimentalCoroutinesApi
+class LoginViewModelTest : ViewModelTest() {
+    private val navigator: Navigator = mockk()
+    private val userRepository: UserRepository = mockk()
+    private val loginUseCase: LoginUseCase = LoginUseCase(testDispatcher, userRepository)
+    private val viewModel: LoginViewModel = LoginViewModel(navigator, loginUseCase)
+
+    @Test
+    fun `username change each time user types new input`() = runTest {
```
Comment (huuphuoc1396) [inline app/src/test/java/com/example/tmdb/ui/features/login/LoginViewModelTest.kt:25]: You should check with the password change as well.

File: data/src/androidTest/java/com/example/tmdb/data/storages/datastores/EncryptedPrefsDatastoreImplTest.kt [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import io.kotest.matchers.shouldBe
+import kotlinx.coroutines.test.TestScope
+import kotlinx.coroutines.test.runTest
+import org.junit.Test
+
+class EncryptedPrefsDatastoreImplTest {
+    private val testScope = TestScope()
+    private val context = ApplicationProvider.getApplicationContext<Context>()
+    private val encryptedUserDatastoreImpl = EncryptedPrefsDatastoreImpl(context)
+    private val TOKEN_TYPE = "Bear"
```
Comment (huuphuoc1396) [inline data/src/androidTest/java/com/example/tmdb/data/storages/datastores/EncryptedPrefsDatastoreImplTest.kt:15]: const value should be declared on the companion scope.

Comment (bententuan) [reply data/src/androidTest/java/com/example/tmdb/data/storages/datastores/EncryptedPrefsDatastoreImplTest.kt]: Fixed

File: app/src/test/java/com/example/tmdb/ui/features/login/LoginViewModelTest.kt (hunk 2 above)
Comment (bententuan) [reply app/src/test/java/com/example/tmdb/ui/features/login/LoginViewModelTest.kt:29]: Fixed

File: app/src/test/java/com/example/tmdb/ui/features/login/LoginViewModelTest.kt (hunk 1 above)
Comment (bententuan) [reply app/src/test/java/com/example/tmdb/ui/features/login/LoginViewModelTest.kt:83]: Fixed

**Review by huuphuoc1396 — APPROVED** [review]

**Pull request merged**